### PR TITLE
[test] Accumulated gradient optimization tests

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR make a test slightly stronger.

Fixes # (issue)

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.    
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by justusschock — APPROVED** [review]

**Comment by SeanNaren** [comment]
hey @tchaton what's the goal here? I think I understand how this makes the test clearer but would be good to get a short description here

**Review by teddykoker — APPROVED** [review]
Nice!

**Comment by tchaton** [comment]
> hey @tchaton what's the goal here? I think I understand how this makes the test clearer but would be good to get a short description here

Yes, I will do. I started to work on accumulated_gradient bug and couldn't find the fix yet. However, I extended the test and make sure it was not updating during accumulation steps. So I submitted the next test in standalone.

T.C

**Comment by SeanNaren** [comment]
> > hey @tchaton what's the goal here? I think I understand how this makes the test clearer but would be good to get a short description here
> 
> Yes, I will do. I started to work on accumulated_gradient bug and couldn't find the fix yet. However, I extended the test and make sure it was not updating during accumulation steps. So I submitted the next test in standalone.
> 
> T.C

Thanks Thomas! After reading the code this is a neat clean up of the existing test.

File: tests/trainer/test_trainer.py [hunk 1]
```diff
@@ -305,28 +305,28 @@ def _optimizer_step(
...
-
-            if (batch_idx + 1) == self.trainer.num_training_batches:
-                for key in self.loss_backward.keys():
-                    # exclude the check for batch_norm parameters
-                    if not any([k in key for k in _exclude_keys]):
-                        assert not torch.equal(self.loss_backward[key], self.opt_step[key])
+        def on_batch_end(self, outputs, batch, batch_idx, dataloader_idx):
+            self.on_train_batch_start_end_dict = self.state_dict()
+            for key in self.on_train_batch_start_end_dict.keys():
+                # exclude the check for batch_norm parameters
```
Comment (SeanNaren) [inline tests/trainer/test_trainer.py:316]: Don't think we need this anymore

```suggestion
```

**Review by SeanNaren — APPROVED** [review]
